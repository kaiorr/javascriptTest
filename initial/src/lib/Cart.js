import find from 'lodash/findLastKey'
import remove from 'lodash/remove'

export default class Cart {
  items = []

  add(item) {
    const itemToFind = { product: item.product }

    if(find(this.items, itemToFind)) {
      remove(find(this.items, itemToFind))
    }
    this.items.push(item)
  }

  getTotal() {
    return this.items.reduce((acumulator, item) => {
      return acumulator + item.quantity * item.product.price
    }, 0)
  }
}
