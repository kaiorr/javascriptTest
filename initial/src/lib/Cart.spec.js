import Cart from './Cart'

describe('Cart', () => {
  let cart;
  let product = {
    title: 'Nike Shox shoes - men',
    price: 3588
  }

  let product2 = {
    title: 'Adidas running shoes - women',
    price: 4122
  }

  beforeEach(() => {
    cart = new Cart()
  })

  describe('getTotal()', () => {

  it('should return 0 when getTotal() is executed in a newly created ', () => {
    expect(cart.getTotal()).toEqual(0)
  });

  it('should  multiply quantity and price and receive the total amount', () => {
    const item = {
      product: {
        title: 'Adidas running shoes - men',
        price: 3588
      },
      quantity: 2
    }
    cart.add(item)

    expect(cart.getTotal()).toEqual(7176)
  });

  it('should ensure no more than on product exists at a time ', () => {
    cart.add({
      product,
      quantity: 2
    });

    cart.add({
      product,
      quantity: 1,
    })

    expect(cart.getTotal()).toEqual(10764)
  });

  it('should update total when a product gets included and then remove', () => {
    cart.add({
      product,
      quantity: 2,
    })

    cart.add({
      product: product2,
      quantity: 1
    })

    cart.remove(product)

    expect(cart.getTotal()).toEqual(4122)
  });

  })

  describe('checkout()', () => {
    it('should return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 2
      })

      cart.add({
        product: product2,
        quantity: 3
      })

      expect(cart.checkout()).toMatchSnapshot()
    });
  });

})
