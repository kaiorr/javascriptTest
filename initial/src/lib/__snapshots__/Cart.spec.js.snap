// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 3588,
        "title": "Nike Shox shoes - men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 4122,
        "title": "Adidas running shoes - women",
      },
      "quantity": 3,
    },
  ],
  "total": 19542,
}
`;
